File: Playwright/PROJECTS/PROJECT_17_BDD_Framework/features/steps/navigation_steps.ts
import { When, Then } from '@cucumber/cucumber';
import { PlaywrightWorld } from '../../support/PlaywrightWorld';
import { ProductsPage } from '../../pages/ProductsPage';
import { CartPage } from '../../pages/CartPage';

When('I navigate to the products page', async function (this: PlaywrightWorld) {
  await this.page!.goto(`${this.baseUrl}/inventory.html`);
  await this.page!.waitForTimeout(1000);
});

When('I click the cart icon', async function (this: PlaywrightWorld) {
  const productsPage = new ProductsPage(this.page!);
  await productsPage.clickCartIcon();
  await this.page!.waitForTimeout(1000);
});

When('I click the {string} button in navigation', async function (this: PlaywrightWorld, buttonText: string) {
  if (buttonText === 'Continue Shopping') {
    const cartPage = new CartPage(this.page!);
    await cartPage.clickContinueShopping();
  }
  await this.page!.waitForTimeout(1000);
});

When('I navigate through the application', async function (this: PlaywrightWorld) {
  // Navigate through different pages
  await this.page!.goto(`${this.baseUrl}/inventory.html`);
  await this.page!.waitForTimeout(500);
  const productsPage = new ProductsPage(this.page!);
  await productsPage.clickCartIcon();
  await this.page!.waitForTimeout(1000);
});

Then('I should see the product list', async function (this: PlaywrightWorld) {
  const productsPage = new ProductsPage(this.page!);
  const isLoaded = await productsPage.isLoaded();
  if (!isLoaded) {
    throw new Error('Products list is not visible');
  }
});

Then('I should see at least one product', async function (this: PlaywrightWorld) {
  const productsPage = new ProductsPage(this.page!);
  const productCount = await productsPage.getProductCount();
  if (productCount === 0) {
    throw new Error('No products found');
  }
});

Then('I should be redirected to the cart page', async function (this: PlaywrightWorld) {
  const cartPage = new CartPage(this.page!);
  const isLoaded = await cartPage.isLoaded();
  if (!isLoaded) {
    throw new Error('Cart page is not loaded');
  }
});

Then('the URL should contain {string}', async function (this: PlaywrightWorld, urlPart: string) {
  const currentUrl = this.page!.url();
  if (!currentUrl.includes(urlPart)) {
    throw new Error(`URL does not contain ${urlPart}. Current URL: ${currentUrl}`);
  }
});

Then('I should return to the products page', async function (this: PlaywrightWorld) {
  const productsPage = new ProductsPage(this.page!);
  const isLoaded = await productsPage.isLoaded();
  if (!isLoaded) {
    throw new Error('Products page is not loaded');
  }
});

Then('all menu links should be accessible', async function (this: PlaywrightWorld) {
  // Verify menu links are accessible
  const productsPage = new ProductsPage(this.page!);
  await productsPage.clickMenuButton();
  await this.page!.waitForTimeout(500);
  const menuVisible = await this.page!.locator('.bm-menu').isVisible();
  if (!menuVisible) {
    throw new Error('Menu is not accessible');
  }
});

File: selenium-java-tests/PROJECT_17_BDD_Framework/src/test/java/com/qa/automation/project17/steps/NavigationSteps.java
package com.qa.automation.project17.steps;

import com.qa.automation.project17.base.DriverHolder;
import com.qa.automation.project17.pages.CartPage;
import com.qa.automation.project17.pages.ProductsPage;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.openqa.selenium.WebDriver;

import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Step definitions for navigation.feature (TC-BDD-NAV-001 to 004).
 */
public class NavigationSteps {

    private static WebDriver driver() {
        return DriverHolder.get();
    }

    private static ProductsPage productsPage() {
        return new ProductsPage(driver());
    }

    private static CartPage cartPage() {
        return new CartPage(driver());
    }

    @When("I navigate to the products page.")
    public void i_navigate_to_the_products_page() {
        if (!driver().getCurrentUrl().contains("inventory")) {
            new LogoutSteps().i_am_logged_in();
        }
    }

    @Then("I see the product list and at least one product.")
    public void i_see_the_product_list_and_at_least_one_product() {
        assertTrue(productsPage().isLoaded());
        assertTrue(productsPage().getProductCount() >= 1, "At least one product");
    }

    @When("I click the cart icon.")
    public void i_click_the_cart_icon() {
        productsPage().clickCartIcon();
    }

    @Then("I am on the cart page and URL contains {string}.")
    public void i_am_on_the_cart_page_and_url_contains(String urlPart) {
        assertTrue(cartPage().isLoaded());
        assertTrue(cartPage().getCurrentUrl().contains(urlPart));
    }

    @When("I click {string}.")
    public void i_click_continue_shopping(String buttonText) {
        if (buttonText.equals("Continue Shopping")) {
            cartPage().clickContinueShopping();
        }
    }

    @Then("I return to the products page.")
    public void i_return_to_the_products_page() {
        assertTrue(productsPage().isLoaded());
        assertTrue(driver().getCurrentUrl().contains("inventory"));
    }

    @When("I open the menu.")
    public void i_open_the_menu() {
        productsPage().openMenu();
    }

    @Then("the menu is visible and I can close it.")
    public void the_menu_is_visible_and_i_can_close_it() {
        // Menu was opened in previous step; page has menu (no explicit close needed for this TC)
        assertTrue(productsPage().isLoaded());
    }
}
